spanner/changestreams/src/main/java/com/example/spanner/changestreams/QueryChangeStreamSample.java

```diff
@@ -0,0 +1,175 @@
+      List<ChangeStreamRecord> initialQueryRecords = executeChangeStreamQueryAndPrint(
+          dbClient, changeStreamName, startTimestampStr, endTimestampStr, null);
+
+      System.out.println("Executing change stream partition queries.");
+      for (ChangeStreamRecord record : initialQueryRecords) {
+        // Executes a partition query to print data records that we just inserted.
+        if (record instanceof ChildPartitionsRecord) {
+          for (ChildPartition childPartition :
+              ((ChildPartitionsRecord) record).getChildPartitions()) {
+            executeChangeStreamQueryAndPrint(
+                dbClient, changeStreamName, startTimestampStr, endTimestampStr,
+                childPartition.getToken());
+          }
+        }
+      }
+    } catch (Exception e) {
+      System.out.println(e);
+    } finally {
+      spanner.close();
+    }
+  }
+
+  // Insert two rows into Singers table.
+  static void insertRows(DatabaseClient client, String tableName) {
+    client.write(
+        ImmutableList.of(
+        Mutation.newInsertOrUpdateBuilder(tableName)
+          .set("SingerId").to(1)
+          .set("FirstName").to("Melissa")
+          .set("LastName").to("Garcia")
+          .build(),
+        Mutation.newInsertOrUpdateBuilder(tableName)
+          .set("SingerId").to(2)
+          .set("FirstName").to("Dylan")
+          .set("LastName").to("Shaw")
+          .build()
+      )
+    );
+  }
+
+  // Execute a change stream query, return and print out the result records.
+  static List<ChangeStreamRecord> executeChangeStreamQueryAndPrint(
+      DatabaseClient dbClient, String changeStreamName, String startTimestamp,
+      String endTimestamp, String partitionToken) {
+    System.out.println("Executing a change stream query with: "
+        + "start_timestamp => " + startTimestamp
+        + ", end_timestamp => " + endTimestamp
+        + ", partition_token => " + partitionToken
+        + ", heartbeat_milliseconds => 5000"
+        + ", read_options => null");
+
+    final String query =
+        String.format("SELECT * FROM READ_%s ("
+          + "start_timestamp => @startTimestamp,"
+          + "end_timestamp => @endTimestamp,"
+          + "partition_token => @partitionToken,"
+          + "read_options => null,"
+          + "heartbeat_milliseconds => @heartbeatMillis"
+          + ")", changeStreamName);
+
+    final ResultSet resultSet =
+        dbClient
+          .singleUse()
+          .executeQuery(
+            Statement.newBuilder(query)
+              .bind("startTimestamp").to(Timestamp.parseTimestamp(startTimestamp))
+              .bind("endTimestamp").to(Timestamp.parseTimestamp(endTimestamp))
+              .bind("partitionToken").to(partitionToken)
+              .bind("heartbeatMillis").to(5000)
+              .build());
+
+    ChangeStreamRecordMapper changeStreamRecordMapper = new ChangeStreamRecordMapper();
+    List<ChangeStreamRecord> res = new LinkedList<>();
```

Comment:
It is a bit peculiar we are using a linked list here. Any reason you don't want to use an ArrayList?

Reply:
Changed to ArrayList